helix-core/src/main/java/org/apache/helix/model/Message.java

```diff
@@ -969,23 +987,32 @@ public boolean isValid() {
 
   /**
    * This class is for categorizing state transition messages based on certain properties, and generating
+   * an identifier string for each category
    */
   public static class MessageInfo {
     public String _msgType;
     public String _resourceName;
     public String _fromState;
     public String _toState;
+    public STRebalanceType _sTRebalanceType;
 
+    /**
+     * This class is used for defining what properties are used to categorize messages.
+     * Note that the bases should be arranged in a lower level -> higher level order,
+     * because we do backwards searching when looking for customized threadpool.
+     */
     public enum MessageIdentifierBase {
-      PER_RESOURCE,
-      PER_STATE_TRANSITION_TYPE
+      PER_RESOURCE, // Most basic level (L0), just include message type and resource name
+      PER_REBALANCE_TYPE, // One level above PER_RESOURCE (L1), concatenate rebalance type (load / recovery) after PER_RESOURCE identifier string
```

Comment:
nit: maybe the enum order align with categorize order?

Reply:
I put PER_REBALANCE_TYPE before PER_STATE_TRANSITION_TYPE is because currently most users are using  PER_STATE_TRANSITION_TYPE, so when we do reverse order traversal to look up for the threadpool, we search PER_STATE_TRANSITION_TYPE first, so it saves some checks. Since these two are in the same level, it should be okay.
But this is a good suggestion, I found that the enum and the order in getMessageIdentifier() is not aligned, so I updated it, please take a look, thanks.